IPython/core/oinspect.py

```diff
@@ -541,118 +543,168 @@ def _format_fields(self, fields, title_width=0):
-                       ])
+            append_field(mime, 'Length', 'length')
+            append_field(mime, 'File', 'file'),
+            append_field(mime, 'Signature', 'definition', code_formatter)
 
             # Source or docstring, depending on detail level and whether
             # source found.
-            if detail_level > 0 and info['source'] is not None:
-                displayfields.append(("Source",
-                                      self.format(cast_unicode(info['source']))))
-            elif info['docstring'] is not None:
-                displayfields.append(("Docstring", info["docstring"]))
-
-            add_fields([("Class docstring", "class_docstring"),
-                        ("Init docstring", "init_docstring"),
-                        ("Call signature", "call_def"),
-                        ("Call docstring", "call_docstring")])
-        
-        if displayfields:
-            return self._format_fields(displayfields)
-        else:
-            return u''
-    
+            if detail_level > 0:
+                append_field(mime, 'Source', 'source', code_formatter)
+            else:
+                append_field(mime, 'Docstring', 'docstring', formatter)
+
+            append_field(mime, 'Class docstring', 'class_docstring', formatter)
+            append_field(mime, 'Init docstring', 'init_docstring', formatter)
+            append_field(mime, 'Call signature', 'call_def', code_formatter)
+            append_field(mime, 'Call docstring', 'call_docstring', formatter)
+
+        return mime
+
     def pinfo(self, obj, oname='', formatter=None, info=None, detail_level=0):
         """Show detailed information about an object.
 
         Optional arguments:
 
         - oname: name of the variable pointing to the object.
 
-        - formatter: special formatter for docstrings (see pdoc)
+        - formatter: callable (optional)
+              A special formatter for docstrings.
+
+              The formatter is a callable that takes a string as an input
+              and returns either a formatted string or a mime type bundle
+              in the form of a dictionnary.
+
+              Although the support of custom formatter returning a string
+              instead of a mime type bundle is deprecated.
 
         - info: a structure with some information fields which may have been
           precomputed already.
 
         - detail_level: if set to 1, more information is given.
         """
-        text = self._format_info(obj, oname, formatter, info, detail_level)
-        if text:
-            page.page(text)
-    
+        info = self._get_info(obj, oname, formatter, info, detail_level)
+        if info:
+            page.page(info)
+
     def info(self, obj, oname='', formatter=None, info=None, detail_level=0):
+        """DEPRECATED. Compute a dict with detailed information about an object.
+        """
+        warnings.warn('Inspector.info is deprecated as of IPython 5.0',
```

Comment:
Actually maybe we should keep the original info. Originally I thought that this should not be public API at all.